In [0]:
import pandas as pd
import numpy as np

In [0]:
training = pd.read_excel('Training Data.xlsx')

In [3]:
training.head()

Company Name Exchange:Ticker  \
0                      024 Pharma, Inc. (OTCPK:EEIG)      OTCPK:EEIG   
1            1-800-FLOWERS.COM, Inc. (NasdaqGS:FLWS)   NasdaqGS:FLWS   
2                 12 Retech Corporation (OTCPK:RETC)      OTCPK:RETC   
3  1347 Property Insurance Holdings, Inc. (Nasdaq...    NasdaqGM:PIH   
4                     1847 Holdings LLC (OTCPK:EFSH)      OTCPK:EFSH   

     Company Type Company Status                Geographic Locations  \
0  Public Company      Operating  United States of America (Primary)   
1  Public Company      Operating  United States of America (Primary)   
2  Public Company      Operating  United States of America (Primary)   
3  Public Company      Operating  United States of America (Primary)   
4  Public Company      Operating  United States of America (Primary)   

                                Business Description  \
0  024 Pharma, Inc. provides healthcare products ...   
1  1-800-Flowers.com, Inc., together with its sub...   
2  12 Retech Corporation, through its subsidiarie...   
3  1347 Property Insurance Holdings, Inc., throug...   
4  1847 Holdings LLC, through its subsidiaries, p...   

                            Industry Classifications       Security Tickers  
0  Beauty Care Products (Primary); Consumer Stapl...             OTCPK:EEIG  
1  Catalog Flowers, Gifts and Novelties (Primary)...  NasdaqGS:FLWS; DB:FWC  
2  Information Technology (Primary); Internet Sof...             OTCPK:RETC  
3  Casualty (Primary); Financials (Primary); Fire...           NasdaqGM:PIH  
4  Commercial and Professional Services (Primary)...             OTCPK:EFSH

In [0]:
comp_desc = pd.ExcelFile('Company_ Business Description.xlsx')

In [0]:
testing = comp_desc.parse(0)

In [6]:
testing.head()

Company Name Exchange:Ticker  \
0                      024 Pharma, Inc. (OTCPK:EEIG)      OTCPK:EEIG   
1            1-800-FLOWERS.COM, Inc. (NasdaqGS:FLWS)   NasdaqGS:FLWS   
2                 12 Retech Corporation (OTCPK:RETC)      OTCPK:RETC   
3  1347 Property Insurance Holdings, Inc. (Nasdaq...    NasdaqGM:PIH   
4                     1847 Holdings LLC (OTCPK:EFSH)      OTCPK:EFSH   

     Company Type Company Status                Geographic Locations  \
0  Public Company      Operating  United States of America (Primary)   
1  Public Company      Operating  United States of America (Primary)   
2  Public Company      Operating  United States of America (Primary)   
3  Public Company      Operating  United States of America (Primary)   
4  Public Company      Operating  United States of America (Primary)   

                                Business Description       Security Tickers  
0  024 Pharma, Inc. provides healthcare products ...             OTCPK:EEIG  
1  1-800-Flowers.com, Inc., together with its sub...  NasdaqGS:FLWS; DB:FWC  
2  12 Retech Corporation, through its subsidiarie...             OTCPK:RETC  
3  1347 Property Insurance Holdings, Inc., throug...           NasdaqGM:PIH  
4  1847 Holdings LLC, through its subsidiaries, p...             OTCPK:EFSH

In [7]:
testing.head()

Company Name Exchange:Ticker  \
0                      024 Pharma, Inc. (OTCPK:EEIG)      OTCPK:EEIG   
1            1-800-FLOWERS.COM, Inc. (NasdaqGS:FLWS)   NasdaqGS:FLWS   
2                 12 Retech Corporation (OTCPK:RETC)      OTCPK:RETC   
3  1347 Property Insurance Holdings, Inc. (Nasdaq...    NasdaqGM:PIH   
4                     1847 Holdings LLC (OTCPK:EFSH)      OTCPK:EFSH   

     Company Type Company Status                Geographic Locations  \
0  Public Company      Operating  United States of America (Primary)   
1  Public Company      Operating  United States of America (Primary)   
2  Public Company      Operating  United States of America (Primary)   
3  Public Company      Operating  United States of America (Primary)   
4  Public Company      Operating  United States of America (Primary)   

                                Business Description       Security Tickers  
0  024 Pharma, Inc. provides healthcare products ...             OTCPK:EEIG  
1  1-800-Flowers.com, Inc., together with its sub...  NasdaqGS:FLWS; DB:FWC  
2  12 Retech Corporation, through its subsidiarie...             OTCPK:RETC  
3  1347 Property Insurance Holdings, Inc., throug...           NasdaqGM:PIH  
4  1847 Holdings LLC, through its subsidiaries, p...             OTCPK:EFSH

In [0]:
Industry = training['Industry Classifications']

Here, I will just take the 1st entries(as per the instruction ) of each row in Industry Classifications

In [0]:
Clean_Class = []

In [0]:
for i in range(len(Industry)):
  temp = Industry[i].split(';')[0]
  Clean_Class.append(temp)

In [0]:
training['Industry Classifications'] = Clean_Class

In [12]:
training.head()

Company Name Exchange:Ticker  \
0                      024 Pharma, Inc. (OTCPK:EEIG)      OTCPK:EEIG   
1            1-800-FLOWERS.COM, Inc. (NasdaqGS:FLWS)   NasdaqGS:FLWS   
2                 12 Retech Corporation (OTCPK:RETC)      OTCPK:RETC   
3  1347 Property Insurance Holdings, Inc. (Nasdaq...    NasdaqGM:PIH   
4                     1847 Holdings LLC (OTCPK:EFSH)      OTCPK:EFSH   

     Company Type Company Status                Geographic Locations  \
0  Public Company      Operating  United States of America (Primary)   
1  Public Company      Operating  United States of America (Primary)   
2  Public Company      Operating  United States of America (Primary)   
3  Public Company      Operating  United States of America (Primary)   
4  Public Company      Operating  United States of America (Primary)   

                                Business Description  \
0  024 Pharma, Inc. provides healthcare products ...   
1  1-800-Flowers.com, Inc., together with its sub...   
2  12 Retech Corporation, through its subsidiarie...   
3  1347 Property Insurance Holdings, Inc., throug...   
4  1847 Holdings LLC, through its subsidiaries, p...   

                         Industry Classifications       Security Tickers  
0                  Beauty Care Products (Primary)             OTCPK:EEIG  
1  Catalog Flowers, Gifts and Novelties (Primary)  NasdaqGS:FLWS; DB:FWC  
2                Information Technology (Primary)             OTCPK:RETC  
3                              Casualty (Primary)           NasdaqGM:PIH  
4  Commercial and Professional Services (Primary)             OTCPK:EFSH

In [13]:
len(set(training['Geographic Locations'])),  len(set(testing['Geographic Locations']))

(1, 2)

In [14]:
set(testing['Geographic Locations'])

{'United States of America (Primary)',
 'United States of America (Primary); United States of America (Primary)'}

Basically, Geographic Location will being a unique value will not affect the prediction.

In [0]:
training.drop('Geographic Locations', axis=1, inplace=True)

In [0]:
 testing.drop('Geographic Locations', axis=1, inplace=True)

There is a unique Company Status (Operating). SO this will also not affect the prediction

In [0]:
training.drop('Company Status', axis=1, inplace=True)

In [0]:
testing.drop('Company Status' , axis=1, inplace=True)

In [23]:
len(set(training['Company Type'])),  len(set(testing['Company Type']))

(1, 1)

In [0]:
training.drop('Company Type', axis=1, inplace=True)

In [0]:
testing.drop('Company Type', axis=1, inplace=True )

In [58]:
training.drop('Company Name', axis=1, inplace=True), testing.drop('Company Name', axis=1, inplace=True)

(None, None)

In [0]:
training.head()

Exchange:Ticker                               Business Description  \
0      OTCPK:EEIG  024 Pharma, Inc. provides healthcare products ...   
1   NasdaqGS:FLWS  1-800-Flowers.com, Inc., together with its sub...   
2      OTCPK:RETC  12 Retech Corporation, through its subsidiarie...   
3    NasdaqGM:PIH  1347 Property Insurance Holdings, Inc., throug...   
4      OTCPK:EFSH  1847 Holdings LLC, through its subsidiaries, p...   

                         Industry Classifications       Security Tickers  
0                  Beauty Care Products (Primary)             OTCPK:EEIG  
1  Catalog Flowers, Gifts and Novelties (Primary)  NasdaqGS:FLWS; DB:FWC  
2                Information Technology (Primary)             OTCPK:RETC  
3                              Casualty (Primary)           NasdaqGM:PIH  
4  Commercial and Professional Services (Primary)             OTCPK:EFSH

In [28]:
testing.head()

Company Name Exchange:Ticker  \
0                      024 Pharma, Inc. (OTCPK:EEIG)      OTCPK:EEIG   
1            1-800-FLOWERS.COM, Inc. (NasdaqGS:FLWS)   NasdaqGS:FLWS   
2                 12 Retech Corporation (OTCPK:RETC)      OTCPK:RETC   
3  1347 Property Insurance Holdings, Inc. (Nasdaq...    NasdaqGM:PIH   
4                     1847 Holdings LLC (OTCPK:EFSH)      OTCPK:EFSH   

                                Business Description       Security Tickers  
0  024 Pharma, Inc. provides healthcare products ...             OTCPK:EEIG  
1  1-800-Flowers.com, Inc., together with its sub...  NasdaqGS:FLWS; DB:FWC  
2  12 Retech Corporation, through its subsidiarie...             OTCPK:RETC  
3  1347 Property Insurance Holdings, Inc., throug...           NasdaqGM:PIH  
4  1847 Holdings LLC, through its subsidiaries, p...             OTCPK:EFSH

In [29]:
len(set(training['Exchange:Ticker'])), len(set(testing['Exchange:Ticker'])), len(set(training['Security Tickers'])),len(set(testing['Security Tickers']))

(1991, 8388, 1996, 8402)

Such similarity in the unique number of Tickers and total number of rows for both training and testing datasets clearly depicts that these columns are of little use for modelling. They are unique ID for company

In [30]:
training.drop(['Exchange:Ticker', 'Security Tickers'], axis=1, inplace=True), testing.drop(['Exchange:Ticker', 'Security Tickers'], axis=1, inplace=True)

(None, None)

In [0]:
import nltk, spacy

In [34]:
pip install spacy && python -m spacy download en_core_web_md

    100% |████████████████████████████████| 17.3MB 1.9MB/s 
featuretools 0.4.1 has requirement pandas>=0.23.0, but you'll have pandas 0.22.0 which is incompatible.
datascience 0.10.6 has requirement folium==0.2.1, but you'll have folium 0.8.3 which is incompatible.
albumentations 0.1.12 has requirement imgaug<0.2.7,>=0.2.5, but you'll have imgaug 0.2.8 which is incompatible.
  Found existing installation: numpy 1.14.6
    Uninstalling numpy-1.14.6:
      Successfully uninstalled numpy-1.14.6
    100% |████████████████████████████████| 120.9MB 2.1MB/s 
  Running setup.py install for en-core-web-md ... done

    Linking successful
    /usr/local/lib/python3.6/dist-packages/en_core_web_md -->
    /usr/local/lib/python3.6/dist-packages/spacy/data/en_core_web_md

    You can now load the model via spacy.load('en_core_web_md')



In [35]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [0]:
from nltk.tokenize.toktok import ToktokTokenizer
import unicodedata

In [0]:
import re

In [0]:
tokenizer = ToktokTokenizer()
stopword_list = nltk.corpus.stopwords.words('english')


In [0]:
nlp = spacy.load('en_core_web_md')

Now, Special Characters and Numbers need to be removed.

In [0]:
def remove_special_characters(text, remove_digits=False):
    pattern = r'[^a-zA-z0-9\s]' if not remove_digits else r'[^a-zA-z\s]'
    text = re.sub(pattern, '', text)
    return text

In [41]:
remove_special_characters("Well this was fun! What do you think? 123#@!", remove_digits=True) # Checking the implementation of the function.

'Well this was fun What do you think '

It should be better if the words are "Stemmed" to their roots.

In [0]:
def stemmer(text):
    ps = nltk.porter.PorterStemmer()
    text = ' '.join([ps.stem(word) for word in text.split()])
    return text

In [44]:
stemmer("My system keeps crashing his crashed yesterday, ours crashes daily")

'My system keep crash hi crash yesterday, our crash daili'

In [0]:
def lemmatize_text(text):
    text = nlp(text)
    text = ' '.join([word.lemma_ if word.lemma_ != '-PRON-' else word.text for word in text])
    return text

In [46]:
lemmatize_text("My system keeps crashing! his crashed yesterday, ours crashes daily") #Checking the implementation of my function

'My system keep crash ! his crash yesterday , ours crash daily'

I think lemmatizing is better than Stemming in our case

Stopwords can be removed for better efficiency of the model that can be used for prediction.

In [0]:
def remove_stopwords(text, is_lower_case=False):
    tokens = tokenizer.tokenize(text)
    tokens = [token.strip() for token in tokens]
    if is_lower_case:
        filtered_tokens = [token for token in tokens if token not in stopword_list]
    else:
        filtered_tokens = [token for token in tokens if token.lower() not in stopword_list]
    filtered_text = ' '.join(filtered_tokens)    
    return filtered_text


In [50]:
remove_stopwords("The, and, if are stopwords, computer is not") #checking the implementation of my function

', , stopwords , computer'

Now, combining the preprocessing functions into a single function

In [0]:
def Preprocess(doc, text_lower_case=True, text_lemmatization=True, special_char_removal=True,stopword_removal=True):    
    #for doc in corpus:
        #lowecasing the sentence(s)        
    if text_lower_case:
        doc = doc.lower()
        # lemmatizing the text
    if text_lemmatization:
        doc = lemmatize_text(doc)
        # removing the  special characters and digits    
    if special_char_removal:
            # insert spaces between special characters to isolate them    
        doc = remove_special_characters(doc, remove_digits=True)  
        # removing the stopwords
    if stopword_removal:
        doc = remove_stopwords(doc, is_lower_case=text_lower_case)
            
           
    return doc

In [0]:
training1= training

In [0]:
training1 = training1['Business Description'].apply(Preprocess)

In [0]:
training['Business Description'] = training1

In [62]:
training.head()

Business Description  \
0  pharma inc provide healthcare product worldwid...   
1  flowerscom inc together subsidiary provide gou...   
2  retech corporation subsidiary operate integrat...   
3  property insurance hold inc subsidiary provide...   
4  hold llc subsidiary provide range product serv...   

                         Industry Classifications  
0                  Beauty Care Products (Primary)  
1  Catalog Flowers, Gifts and Novelties (Primary)  
2                Information Technology (Primary)  
3                              Casualty (Primary)  
4  Commercial and Professional Services (Primary)

In [0]:
testing1 = testing['Business Description'].apply(Preprocess)

In [0]:
testing['Business Description']=testing1

In [65]:
testing.head()

Business Description
0  pharma inc provide healthcare product worldwid...
1  flowerscom inc together subsidiary provide gou...
2  retech corporation subsidiary operate integrat...
3  property insurance holding inc subsidiary prov...
4  holding llc subsidiary provide range product s...

Now, both the training and testing datasets have been preprocessed.

In [66]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [0]:
lem_corpus_words = {}
class_words = {}
classes = list(set(training['Industry Classifications']))
for c in classes:
    # prepare a list of words within each class
    class_words[c] = []

# loop through each sentence in our training data
for i in range(len(training)):
    # tokenize each sentence into words
    for word in nltk.word_tokenize(training.iloc[i]['Business Description']):
              
        if word not in lem_corpus_words:
            lem_corpus_words[word] = 1
        else:
            lem_corpus_words[word] += 1
                
        class_words[training.iloc[i]['Industry Classifications']].extend([word])
            
          



A function to compute the industry class scores for each classes.

In [0]:
def calculate_ind_score(sentence, class_name):
    score = 0
    # tokenize each word in our new sentence
    for word in nltk.word_tokenize(sentence):
        # check to see if the stem of the word is in any of our classes
        if word in class_words[class_name]:
            # treat each word with relative weight
            score += (1 / lem_corpus_words[word])
      
    return score

Final function for predicting the classes based on maximum class scores

In [0]:
def pred_ind_class(df):    
    Final_pred=[]
    for sentence in df['Business Description']:
        pred_Class=[]
        pred_score=[]
        for c in class_words.keys():
            pred_Class.append(c)
            pred_score.append(calculate_ind_score(sentence,c))
        Final_pred.append(pred_Class[np.argmax(pred_score)])
    
    return Final_pred 

In [0]:
pred_train = pred_ind_class(training)

In [85]:
(np.sum(pred_train == training['Industry Classifications'])) / len(training)

0.984015984015984

98.4% accuracy seems to be a good bar for testing. There are over 200 classes and a model acheiving 98% for such classification may be considered a good one.

In [0]:
pred_test = pred_ind_class(testing)

In [0]:
final_business_description = comp_desc.parse(0)

In [0]:
final_business_description['Industry Classifications']= pred_test

In [0]:
final_business_description.to_csv('Final_outcome.csv')